TEP-0154 Concise Remote Resolver Syntax

chitrangpatel opened the pull request:
This PR proposes to include a concise remote resolver syntax to reference remote resources.
It addresses the feature request https://github.com/tektoncd/pipeline/issues/7508.

/kind tep

chitrangpatel commented:
/assign @wlynch

chitrangpatel commented:
cc @chmouel PTAL (specifically for `http-resolver`) 🙂

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,290 @@
+  resolver: "bundle:docker.io/ptasci67/example-oci@sha256:053a6cb9f3711d4527dd0d37ac610e8727ec0288a898d5dfbd79b25bcaa29828?kind=task&name=hello-world?secret=default"
+```
+
+Occasionally, users may want to fetch from a private registry (e.g. at `10.96.190.208:5000`). This host contains a `:` and violates the requirement of a purl [`type`](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#rules-for-each-purl-component) and would lead to an invalid `purl`. In cases where purl is violated, users cannot use the purl `resolver-reference`. They ne ...
````
This is a pretty significant draw back - I think this needs to be in scope.

(same w/ http)

chitrangpatel replied:
Sounds good! Fixed.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,290 @@
+http:<schemaless-url>?http-username=<http-username>&http-password-secret=<http-password-secret>&http-password-secret-key=<http-password-secret-key>
+```
+
+In this case, the http-resolver would first attempt using `https` and then `http` (without TLS).  
````
I don't want to pick on this if it's existing behavior, but this seems problematic 😅 

If the secure way failed we shouldn't fallback to unsecure behavior unless specifically requested.

Let users specify http vs https.

chitrangpatel replied:
This is not the behaviour today. Users provide the protocol they want to use as part of the url. 

I removed this approach. Thanks!

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,290 @@
+Based on `bundle` resolver's [params](https://tekton.dev/docs/pipelines/bundle-resolver/#parameters), we define the following purl spec:
+
+```
+bundle:<bundle-provider>/<bundle-namespace>/<bundle-name>@<bundle-version>?kind=<kind>&name=<name>&secret=<secret>
````
This also gets messy for similar reasons as git. oci spec doesn't make a restriction on path lengths - this can be 1...many parts long.

chitrangpatel replied:
Agreed!  I think its fixed now.

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,290 @@
+
+[Package-URL](https://github.com/package-url/purl-spec) is a format that we have begun to use in Tekton Artifacts. To enable a consistent concise format for across resolvers, we propose using purl.
+
+To provide a consistent format across resolvers, we suggest the following `remote-resolver` template: `<resolver-name>:<type>/<namespace>/<name>@<version>#<subpath>?<additional-params>`.
```
agreed. Fixed accordingly but with slight updates. PTAL

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,290 @@
+Based on `git` resolver's [params](https://tekton.dev/docs/pipelines/git-resolver/#parameters), we define the following purl spec:
+
+```
+git:<git-provider>/<org>/<repo>@<revision>#<pathInRepo>?token=<token>&tokenKey=<tokenKey>&serverURL=<serverURL>&scmType=<scmType>
````
Agreed.

wlynch reviewed (commented):
/lgtm

IIUC we loosened the constraints enough that tl;dr this is adding `url` to the remote resolver request.

dibyom commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,336 @@
+Based on `http` resolver's [params](https://tekton.dev/docs/pipelines/http-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+<http-url>?http-username=<http-username>&http-password-secret=<http-password-secret>&http-password-secret-key=<http-password-secret-key>
````
isn't this too specific to k8s secrets?

chitrangpatel replied:
Those are the currently chosen parameter names for the http resolver:  https://tekton.dev/docs/pipelines/http-resolver/#parameters 

I just used them here for parity. We can probably change them since it is still alpha. But I think it should be consistent and is probably outside the scope of this TEP? (as in we update the names if desired separately)

chmouel replied:
@dibyom isn't that what we want ? being able to pass secrets and key location ? most tekton tasks (in catalog) have that same sort of pattern..

@chitrangpatel all good by me, just wondering how would you handle non url parseable characters, maybe we need to specifyc that they should be url encoded ?

chitrangpatel replied:
Great point @chmouel. 
I will add it.

chitrangpatel replied:
added the point that the url must be url-encoded.

dibyom commented:
/approve
/hold

chmouel commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,336 @@
+```
+<type>/<catalog>/<kind>/<name>@<version>
+```
+**Note** here, the resolver does not work with a `scheme` and `location = <type>/<catalog>/<kind>/<name>`.
````
is there any reasons why this could not be a scheme ? is it not supported by the purl spec?

chitrangpatel replied:
Fixed. I added `hub` as a scheme

vdemeester reviewed (commented):
If we want to be consice, couldn't we use `name` for this instead adding `url` and duplicating some information (the name) ? I think it would be backward compatible : it's `{name}://…` we infer the name and we don't read the rest, it's `{name}` only it's the current syntax that applies. In `ResolutionRequest` we would pass the url or the parameters and let the resolver handle that (and/or the helper code we provide).

Can we write an example for a custom resolver ? They are part of the resolver "feature set" (to be able to write your own resolver) and thus we should make sure we take them into account in this proposal.

chitrangpatel commented:
> If we want to be consice, couldn't we use `name` for this instead adding `url` and duplicating some information (the name) ? I think it would be backward compatible : it's `{name}://…` we infer the name and we don't read the rest, it's `{name}` only it's the current syntax that applies. In `ResolutionRequest` we would pass the url or the parameters and let the resolver handle that (and/or the helper code we provide).
> 
Just the name alone won't help. That's because it then requires that the value starts with the resolver name. And that does not work with something like the http resolver where the scheme could be `http` or `https`. 
e.g.
`https://github.com/tektoncd/pipeline` is probably meant to go to the git resolver.
`https://raw.githubusercontent.com/tektoncd/pipeline/main/examples/v1/taskruns/array-default.yaml` is meant to go to the http resolver.

So in the above case, the name alone does not work. But name with the resolver works because we know where the resolution request is supposed to go:
```
taskRef:
  name: "https://github.com/tektoncd/pipeline"
  resolver: git
```


> Can we write an example for a custom resolver ? They are part of the resolver "feature set" (to be able to write your own resolver) and thus we should make sure we take them into account in this proposal.

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,340 @@
+artifact/tasks-catalog/tasks/task@v1.0 # e.g. for the hub resolver where there is no need for a scheme
+```
+
+#### Git Resolver
````
I think that's up to the resolver to define. So yes, we could choose to add that here. The question is how would the resolver tell the difference between the different cases.

vdemeester replied:
that's where the "problem" may lie. In my head `github://…` is a different resolver than `git+https` or `https`, …
I commented a bit about that in https://github.com/tektoncd/pipeline/issues/7508 😬

chitrangpatel replied:
To me `github://`, `git+https://`, `https` are schemes. I don't think a single resolver has to only be able to handle one scheme. For example, the git resolver could handle:
* https
* git+https
* git+ssh

If an unsupported scheme is passed to the resolver, it can throw an error. How the url is used is also decided by the resolver.
The resolver could choose to use the url as is to handle the resolution request or it could choose to parse and extract the necessary information to construct a different url altogether when performing the resolution. This will partly be dictated by the libraries that the underlying resolver uses. 

Effectively, we want to pass it a string value. We are saying that the best practice is to format that string uniformly using the prescribed template. But how that value is used is determined by the resolver.

chitrangpatel replied:
It is completely ok for the resolver to be able to parse multiple formats:

e.g. 
if scheme == `https`, 
Then it must look like
```
https://<complete-git-repo>@<version/sha>#<sub path>

e.g.

https://github.com/tektoncd/pipeline@XXXXX#a/b/c
```

But if scheme == `scmType`
It must look like:
```
<scmType>://<org>/<repo>@<version>#<sub path>?<tokens etc.>

e.g.
github://tektoncd/pipeline@sha1:nsjdhjshjdh#examples/v1/taskruns/hello.yaml
```

The user can provide either format they choose. And as long as the resolver can parse the string, it's all good.

If users make a mistake while providing the string, the resolver can throw a validation error.

chitrangpatel replied:
May be I can make this a bit more explicit with examples and scenarios.

vdemeester replied:
Yes, I agree with all that you said I think.

chitrangpatel replied:
Done.

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,340 @@
+
+## Future Work
+
+The current remote resolution requires the resolver name to be provided by the Task/Pipeline author. This means that if the same pipeline/Task is applied on another cluster where the resolver names are different then they would not work. It would be nice to make the `resolver-name` optional and let Tekton figure out how to route through the resolvers on the cluster and pass the `url`. This curren ...
```
I'll let @wlynch comment but the idea here was that different resolvers subscribe to schemes (e.g. https, git+https, etc.). The Tekton pipelinerun/taskrun controller parses the scheme and then forwards the request to the resolvers that have subscribed to this scheme one at a time based on some form of priority. 

e.g. a `url` of `https://github.com/tektoncd/pipeline` would first be tried by the git-resolver and if that fails, we would send it to the http resolver because both those resolvers have subscribed to the `https` scheme.

Because something like this involves significant change in the existing architecture, we scoped it out of this TEP into future work. 

So today, the request for:
```
taskRef:
  resolver: git
  url: https://github.com/tektoncd/pipeline
```
Would go explicitly to the git resolver. Once we have this future work, the `resolver: git` would be optional and Tekton would try forwarding the requests.

vdemeester replied:
I like the idea of "scheme subscription" and priority (we'll have to be very careful) 👼🏼 But I would definitely include it in that proposal. It does require some significant change in the architecture, but we'll have to do it at some point, so let's describe here what we want long-term and then let's figure out how we go there.

chmouel replied:
I like that idea as well big +1 on making it more easy for end users to don't care about the concept of resolvers and having to choose one...

chitrangpatel replied:
Ok. I can expand on it a bit more and try to lay out a plan of how we can get there. I didn't want to go into lots of details since they are still TBD and beyond the scope of this TEP.

vdemeester replied:
I would really love to expand the scope of this TEP to include them 👼🏼

chitrangpatel replied:
The reason I put the scheme subscription based solution out of scope for this TEP is because it significantly increases the complexity and the timeline to deliver it. I don't have the capacity to drill into that right now and so I didn't want to expand the scope of I know that I cant deliver on it.

However, I wanted to convince you that the current TEP will get us half way there towards the subscription based approach. 

 The format of the `name/url` (whatever we call it) starts with a scheme. 
 For now, we can't tell based on the scheme as to which resolver to try. That's why we want to pair it with the existing `resolver` field which informs us which resolver to send the `name/url` to. 
 In the future, we can make the `resolver` field optional when we have the scheme subscription based approach implemented. Even then, if a user passes the `resolver` field, then we know exactly which resolver to pass the request to so it is backwards compatible. 
```yaml
 taskRef:
   name: "https://github.com/tektoncd/pipeline@main#examples/v1/taskrun/hello.yaml"
   resolver: git # in the future this is optional.
```

chitrangpatel replied:
However, if others were interested in collaborating with me on this, then its definitely doable.

chitrangpatel replied:
We can also do this in two stages where in stage 1, we do what's currently suggested and stage 2, we implement the scheme based subscription.

chitrangpatel replied:
Resolving this now since its now in the main scope.

vdemeester commented:
@chitrangpatel, yeah, as you commented [here](https://github.com/tektoncd/community/pull/1138#discussion_r1530557633), we would want "resolver" to be able to say "I handle those sheme", and some sort of priority.

Now, even with that, the `git+https` is a slight… tricky one to achieve. If the `http` resolver has a higher priority than `git`, would we always let the `http` resolver handle it, fail, and give it to the next ? In case of an 401 (unauthorized), would we still do the same (with a high level of chance it will be 401 with the git provider as well).

Some history note (that might not have any "impact"), the current syntax is also there because it was decided to be consistent with the rest of the API (on params). The initial proposal looked like

```yaml
  name: …
  params:
    - foo: bar
    - {name}: {value}
```

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,340 @@
+
+[Package-URL](https://github.com/package-url/purl-spec) is a format that we have begun to use in Tekton Artifacts. To enable a consistent concise format for across resolvers, we propose taking purl as an inspiration to define our format. The reason for not using purl directly is that it is too rigid for the different types of resolvers.
+
+To provide a consistent format across resolvers, we suggest the following `remote-resolver` template: `<scheme>://<location>[@<version>][#<selector>][?<params>]`.
```
Yes!

chitrangpatel replied:
I added some more sentences to make that explicit.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,308 @@
+Based on `cluster` resolver's [params](https://tekton.dev/docs/pipelines/cluster-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+cluster://<kind>/<namespace>/<name>
````
nit: Follow k8s resource URI conventions: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-uris

e.g. `NAMESPACE/RESOURCETYPE/NAME`

similar if you need the full GVK, `GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME`

chitrangpatel replied:
Agreed! I will fix this.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,308 @@
+Example:
+
+```
+hub://artifact/tekton-catalog-tasks/task/git-clone@0.6
````
non-blocking, but this can also be

`artifacthub://...` and `tektonhub://...`

chitrangpatel replied:
Yes! I actually like that more.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,308 @@
+ 
+```yaml
+apiVersion: v1
+kind: ConfigMap
````
I actually am not sure whether we want a configmap for this. we probably need a concept of ordered priority, which we can't easily do with simple key-value pairs (and when parsed go maps can make this ordering non-deterministic).

chitrangpatel replied:
I discuss this in the next section but my idea here was that since the value of the keys is a comma-separated list of resolver names, the order is high to low. So the cluster operators can choose the order of priority based on each scheme.

e.g. say we have a cluster with three resolvers: `resolver-a, resolver-b, resolver-c` and across the resolvers, the supported schemes are: `http, https, oci, git+https`. The config would look like:

```yaml
data:
  https: resolver-a, resolver-b
  http: resolver-b,resolver-a
  oci: resolver-c
  git+https: resolver-a
```

Based on above, config, for the scheme `https`, the cluster operator has decided to provide a higher priority to `resolver-a` over `resolver-b` resolver and vice-versa for the scheme `http`.

vdemeester replied:
This approach means the cluster operator needs to install the resolver **and** update some configmap for it to be usable, am I right ?
If that's the case, I am not super fan of the UX (but I am not sure I have something better to propose). In my head, the idea was that resolvers have some default "weight" attached to them (built-in have the highest, so most priority) at build time, and could be overriden / set by operator if they want to change the "default" order, somehow ?

chitrangpatel replied:
Hmm. Good point. 

For built-in resolvers, this is easy and we can order them with some default priority like you suggested.

If users have custom resolvers on the cluster, how would Tekton controller know if that? I thought that a config map like this would help in that case. Even if it can identify other them on the cluster, how does it know about the exact name (what is provided by the `GetName`) of the resolver? In other words, what would the TaskRun/PipelineRun controller set as the label: https://tekton.dev/docs/pipelines/how-to-write-a-resolver/#the-getselector-method?

chitrangpatel replied:
I updated this section and the format of the configMap. PTAL at the design for the updated format. I think with this approach, cluster operators don't need to do anything. When the framework calls the `Initialize` method, the resolvers register their schemes to the configMap that Tekton reads from.

wlynch replied:
We definitely need some sort of data source to configure these - I just don't know if a ConfigMap is the right tool. 🤔 

If multiple resolvers subscribe to a single scheme type it's ambiguous what order that will happen in since ConfigMap key ordering is not guaranteed to remain stable after parsing due to the nature of Go maps. So we'd either need to write our own ConfigMap parser, use a different structure, or use some clever string encoding (though idk if this will be the most readable).

The alternative format you put in https://github.com/tektoncd/community/pull/1138#discussion_r1536674203 is less ambiguous since resolver ordering will be retained in the value. We should firm up the order of operations in particular for `*` (my guess is this should be a catch-all and evaluated last). The trade-off here is that it's more of a pain to add a new resolver - you can't add a new resolver simply by patching in a line addition.

I don't think we should grant configmap rbac to resolvers, though I like that you're thinking about the UX around managing resolvers!

chitrangpatel replied:
I've updated the design based on our discussions. PTAL again.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,308 @@
+  name: remote-resolution-scheme-subscription
+  namespace: tekton-pipelines
+data:
+  #<scheme>: "<resolvers subscribed to the scheme; comma separated>"
```
For the resolver I was hacking on, I wanted to be able to subscribe to all remote resolution requests and handle the url resolving resolver side. How should this be represented here? 🤔

chitrangpatel replied:
Good question. My guess is that in the configMap, we could have a special key: `*` to say that the resolvers that subscribe to all formats.

```yaml
data: 
  *: resolver-billy
  https: resolver-a, resolver-b 
  http: resolver-b,resolver-a 
  oci: resolver-c 
  git+https: resolver-a
```

What's the use case for this? 
I think in general that this case might be a bit tricky because I can't think of a resolver (universal-resolver ) that can resolve any arbitrary scheme. That's because some schemes like for the `hub/cluster` resolver above are made-up schemes that are only understood by them. It is likely meaningless to other resolvers. 

If the resolver can handle multiple schemes, we simply list them next to the the schemes they support like above for `resolver-a`.

chitrangpatel replied:
I've updated the TEP to reflect this better.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,308 @@
+  resolver: "cluster"
+  name: "cluster://task/default/git-clone" 
+```
+
````
Before we get into stage 2, in stage 1 we have multiple resolvers potentially resolving to the same scheme (`http`). What changes (if any) need to be made to the resolution api to support this?

How should a resolver let the controller know that it isn't capable of fulfilling a request?

How is priority determined without phase 2 in place?

chitrangpatel replied:
In phase 1, the `resolver` field is required. So users need to specify the `resolver` and the `name`. This way, the controller knows exactly which resolver to forward the request to. There is no need for priorities or retries. The behaviour is similar to what we have today.

 In phase 2, the `resolver` field  would become **optional** and based soon `name` solely, the controller needs to figure out which resolver to forward the request to. Thats when we need to worry about priority.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,308 @@
+
+### Updating the `name` field
+* Currently the value of the `name` field must be a sub-domain name. We will loosen this restricitons to also allow for a url-like name. 
+* The string field `name` will also be added to the `ResolutionRequest CRD` which will contain the value (after variable replacement) passed without any validation from `taskRef, pipelineRef and ref`.
```
I thought this section was talking about the existing name field in the CRD. 😅 

Since `name` is generic, can you link out to the specific types you want to modify here?

chitrangpatel replied:
Yes, it's a bit confusing. I will clarify it.

We have a `name` in the `TaskRef`, `PipelineRef` and `Ref` (StepAction's):
```
taskRef:
  - name: blah (currently to resolver locally applied tasks)
```
I think this confusion will go away if we add a new field `url` like you suggested above instead of repurposing the field `name`.

wlynch commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,308 @@
+The feature will be gated behind a feature flag: `enable-concise-resolver-syntax`. The feature will be propagated through the lifecycle of `alpha --> beta --> stable` like we do with other features.
+
+### Updating the `name` field
+* Currently the value of the `name` field must be a sub-domain name. We will loosen this restricitons to also allow for a url-like name. 
```
Alternatively, should we add a new `url` field?

chitrangpatel replied:
That was part of the original proposal. I'm personally ok with either approach.

Pros of the new field:

We leave `name` dedicated to resolving the locally applied references.

And the combination of `url`, or `resolver and params` take care of remote resolution.

chitrangpatel replied:
cc @vdemeester WDYT?

vdemeester replied:
I think I am ok with it 😬

chitrangpatel replied:
ok with name or url?

wlynch reviewed (commented):
Overall looks good! just some clarifiying questions

chitrangpatel commented:
> If we want to be consice, couldn't we use `name` for this instead adding `url` and duplicating some information (the name) ? I think it would be backward compatible : it's `{name}://…` we infer the name and we don't read the rest, it's `{name}` only it's the current syntax that applies. In `ResolutionRequest` we would pass the url or the parameters and let the resolver handle that (and/or the helper code we provide).
> 
> Can we write an example for a custom resolver ? They are part of the resolver "feature set" (to be able to write your own resolver) and thus we should make sure we take them into account in this proposal.

Done! I added a section on Custom Resolver. Also, in the Design Details, I added a section on expanding the Resolver interface  so that custom Resolvers can leverage it.

vdemeester commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,326 @@
+<scheme>://<location>[@<version>][#<selector>][?<params>]
+```
+
+The custom resolver can choose the schemes it wants to subscribe to and for each scheme, it can define the remainder of `location, version, selectors and params` as needed. An example `scheme` could be `<custom-resolver-name>`. A `location` may be made up of concrete units e.g. `location = <foo>/<bar>` (i.e. expect two parts) that are up to the discretion of the custom resolvers.
````
The subscribe part applies for stage 2 only though, right ?

chitrangpatel replied:
Yes, let me clarify that.

chitrangpatel replied:
done.

vdemeester commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,326 @@
+
+#### Prioritizing the order of resolvers
+
+Occasionally, for certain common `scheme`s like `https`, multiple resolvers could be subscribed to it. In that case, the Tekton controller would send out requests based on the priority to the list of the subscribed resolvers until one of them succedes in resolving the request. The cluster operators will maintain the priority of the resolvers based on the order of the resolvers subscribed to the s ...
```
nit: the controller could send it to multiple but get the results in order, to optimize time consumption.

chitrangpatel replied:
Great point! It can federate the request to all of them.

This has triggered a thought:
Imagine, we didn't have to worry about a priority/scheme based subscription at all? 
What if we always federated any remote resolution request to all the resolvers? 
In this scenario, most resolvers that are not subscribed to the scheme would immediately fail. Only the ones that understand it would succeed. Is that bad? This would dramatically reduce the complexity.
Right now, the TaskRun/PipelineRun controller identifies which resolvers to forward the request to. Why not have it federate to them all? And then it can round robin and get the resource from the resolver that succeeds. Even if multiple resolvers succeed, there is no harm done, we have what we need.

chitrangpatel replied:
cc @wlynch WDYT?

vdemeester replied:
The downside of this would be, if it works as today, a growing number of objects (each "resolving" would mean a `ResolutionRequest` for each resolver). We tend to have etcd pressure already with the amount of objects we have so…

chitrangpatel replied:
That's a good point. Does the priority matter then? If we send out request to all the subscribed resolvers then as long as we get a response from any of them, that should be fine?

chitrangpatel replied:
I've updated this section. I don't think there is a need for a priority if we federate the request to all the subscribed resolvers simultaneously.

chitrangpatel replied:
It dramatically reduces the complexity since the cluster operators also don't need to maintain configMaps.

chitrangpatel replied:
PTAL at the updated section.

vdemeester commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,326 @@
+
+#### Subscribing to `scheme`s
+
+Each remote resolver will subscribe to `scheme`s. e.g. the git resolver might subscribe to `scheme`s like `git+https`, `https` or a bundle resolver might subscribe to `scheme`s like `oci`.
```
One question that I didn't though of before : what happens if the "other side" of the scheme is different from one provider to the other ? For example, for `https`, with the git resolver, the parameter is `secret`, but for the http resolver it is `http-password-secret` — how should we / the user handle that ?

chitrangpatel replied:
I think if users need to know what parameters they can pass, that is very specific to the resolvers since the names of the parameters are defined by the resolvers. 

In that scenario, users won't be able to try arbitrary parameters. Infact, they will have to know exactly which resolver they are calling and use parameters that the resolver understands.

On our end, we would try both resolvers subscribed to the scheme. The resolvers that do not understand the params from the scheme should error out. The resolvers should do this check when they write `ValidateUrl` that the resolver framework calls.

vdemeester replied:
Fair enough.

vdemeester reviewed (commented):
Thanks for the update @chitrangpatel 🤗

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,326 @@
+
+Currently, users can either use the `name` field or the combination of `resolver` and `params`.
+
+We propose updating the `name` field whose value can also be the [resolver-reference](resolver-reference). This field can be used in addition to the `resolver` instead of mutually exclussive. For example:
```
Yes, thats right. We want to change this behaviour and allow them both.

chitrangpatel replied:
Added more explanation in the TEP.

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,326 @@
+  resolver: git
+```
+
+Alternatively, if the `scheme` is the `scmType` then the format is:
````
That would be tricky because it would be treated as a different `scheme`. May be thats a limitation of this scheme. 
One way I see passing this information is via the previous scheme:
`https://github.my.com/...`

chitrangpatel replied:
Or, in such scenarios, the user can provide the exact resolver they wish to forward the request to so that Tekton does not need to figure out which ones to send the request to.

chitrangpatel replied:
I've reflected this point in the TEP.

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,326 @@
+Where, `scheme` is required while `location`, `version`, `selector` and `params` are all optional and should be parsed by the `Remote Resolvers` according to their needs. Additionally, the value of the `name` must be `url encoded`.
+In [stage 2](#stage-2), remote resolvers will subscribe to `scheme`s that they support and Tekton will auto-forward the resolution request to the subscribed resovers. 
+
+The `Remote Resolver` must define the format using the above `remote-resolver` template. The `Remote Resolver` is be responsible for parsing the `resolver-reference` and extracting the necessary information. The Tekton controller will perform no validation on the value of the `name`. It will simply pass it as is in the `Remote Resolution CRD`. 
```
Yes, great point. Will fix.

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
````diff
@@ -0,0 +1,326 @@
+artifact/tasks-catalog/tasks/task@v1.0 # e.g. for the hub resolver where there is no need for a scheme
+```
+
+##### Git Resolver
````
fixed.

chitrangpatel commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,326 @@
+### Use Cases
+
+* As a task/pipeline author, I would like to reduce the verbosity of the yaml.
+* As a task/pipeline author, I would like my referenced pipeline/task to be runnable on any cluster.
```
I updated it.

vdemeester commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,376 @@
+To provide a consistent format across resolvers, we suggest the following `remote-resolver` template: `<scheme>://<location>[@<version>][#<selector>][?<params>]`. This template is a "best practice" so that all resolvers follow a similar template. It will not be enforced by the Tekton Pipeline/Task controllers. In the template, `scheme` is required while `location`, `version`, `selector` and `para ...
+In [stage 2](#stage-2), remote resolvers will subscribe to `scheme`s that they support and Tekton will auto-forward the resolution request to the subscribed resovers. 
+
+The `Remote Resolver` must define the format using the above `remote-resolver` template. The `Remote Resolver` is be responsible for parsing the `name` and extracting the necessary information. The Tekton controller will only validate that the `name` starts with `<scheme>://`. If a `scheme` is found, it will match against the schemes that the resolvers accept. If provided `scheme` is not supporte ...
```
```suggestion
The `Remote Resolver` must define the format using the above `remote-resolver` template. The `Remote Resolver` is responsible for parsing the `name` and extracting the necessary information. The Tekton controller will only validate that the `name` starts with `<scheme>://`. If a `scheme` is found, it will match against the schemes that the resolvers accept. If provided `scheme` is not supported by any of the resolvers, then it will throw an error. If the `name` does not start with `<scheme>://`, the Tekton controller will then treat it as a local reference and not forward the request to the remote resolvers. Validation of the other optional fields must be conducted by the resolvers themselves. 
```

chitrangpatel replied:
Fixed

vdemeester reviewed (commented):
SGTM 👼🏼

vdemeester commented on teps/0154-concise-remote-resolver-syntax.md:
```diff
@@ -0,0 +1,397 @@
+In this stage, we enable the Tekton controller to be able to identify the resolvers on the clusters that have subscribed to the scheme of the incoming [resolver reference](#resolver-reference).
+
+#### Prioritized Subscription to `scheme`s
+This mapping of `scheme` to `resolver` could be easily maintained in a `configMap` (or a [CRD](#using-crd-to-store-prioritized-subscription-configuration)) for the cluster. For [built-in resolvers](#built-in-resolvers), Tekton installation will be pre-configured with the schemes and priorities so that the cluster operators do not need to worry about any interaction unless they want to update the  ...
```
This will need to be highlighted when we implement this, because it is a change from what happens today (aka, today you install the resolver and you can use it directly).

chitrangpatel replied:
Agreed!

chitrangpatel commented:
/hold cancel

wlynch reviewed (commented):
/lgtm
